Lib/fontTools/varLib/instancer/__init__.py

```diff
@@ -429,7 +429,7 @@ def setMvarDeltas(varfont, deltas):
         if mvarTag not in MVAR_ENTRIES:
             continue
         tableTag, itemName = MVAR_ENTRIES[mvarTag]
-        delta = deltas[rec.VarIdx]
+        delta = deltas.get(rec.VarIdx, 0)
```

Comment:
hm, wouldn't it be better to add a `{NO_VARIATION_INDEX: 0}` before returning `defaultDeltas` from `instantiateItemVariationStore`? That way if `rec.VarIdx` really is no-op it works, but for whatever reason it points to some non existent VarIdx, it would fail with KeyError, this not hiding a bug somewhere in the font or code.
Like this:

```diff
diff --git a/Lib/fontTools/varLib/instancer/__init__.py b/Lib/fontTools/varLib/instancer/__init__.py
index 986ad5a4..91dd586f 100644
--- a/Lib/fontTools/varLib/instancer/__init__.py
+++ b/Lib/fontTools/varLib/instancer/__init__.py
@@ -429,7 +429,7 @@ def setMvarDeltas(varfont, deltas):
         if mvarTag not in MVAR_ENTRIES:
             continue
         tableTag, itemName = MVAR_ENTRIES[mvarTag]
-        delta = deltas.get(rec.VarIdx, 0)
+        delta = deltas[rec.VarIdx]
         if delta != 0:
             setattr(
                 varfont[tableTag],
@@ -633,6 +633,7 @@ def instantiateItemVariationStore(itemVarStore, fvarAxes, axisLimits):
         for major, deltas in enumerate(defaultDeltaArray)
         for minor, delta in enumerate(deltas)
     }
+    defaultDeltas[itemVarStore.NO_VARIATION_INDEX] = 0
     return defaultDeltas


diff --git a/Tests/varLib/instancer/instancer_test.py b/Tests/varLib/instancer/instancer_test.py
index 2757c3da..c3b6f71d 100644
--- a/Tests/varLib/instancer/instancer_test.py
+++ b/Tests/varLib/instancer/instancer_test.py
@@ -456,6 +456,8 @@ class InstantiateItemVariationStoreTest(object):

         defaultDeltaArray = []
         for varidx, delta in sorted(defaultDeltas.items()):
+            if varidx == varStore.NO_VARIATION_INDEX:
+                continue
             major, minor = varidx >> 16, varidx & 0xFFFF
             if major == len(defaultDeltaArray):
                 defaultDeltaArray.append([])

```

Reply:
potentially (even though unlikely) a NO_VARIATION_INDEX could be found inside a GPOS ValueRecord or GDEF Anchor device table. You already modified `VarStoreInstancer.__getitem__` to return 0 when key is NO_VARIATION_INDEX, and this pseudo-dict `instancer` object in used in `MutatorMerger` for instancing GPOS/GDEF, so if a NO_VARIATION_INDEX occurs, it's handled there.

In `varLib.instancer:instantiateOTL` I call `MutatorMerger` with the `defaultDeltas` dict that are returned from `instantiateItemVariationStore` (not with an actual `VarStoreInstancer` object); if this dict doesn't contain a map for {NO_VARIATION_INDEX: 0} it could KeyError.

So my patch above would handle that case as well.

Reply:
Thanks. WIll do.